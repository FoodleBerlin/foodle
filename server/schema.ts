import { makeSchema } from "nexus";
import * as path from "path";
import * as types from "./graphql/types";
import * as PrismaScalars from "nexus-prisma/scalars";

const schema = makeSchema({
  prettierConfig: path.join(process.cwd(), ".prettierrc"),
  types: [types, PrismaScalars.DateTime],
  // TODO implement isTypeOf field in all objects
  // https://nexusjs.org/docs/guides/abstract-types
  features: {
    abstractTypeStrategies: {
      isTypeOf: true,
    },
  },
  nonNullDefaults: {
    output: true,
  },
  outputs: {
    typegen: path.join(
      process.cwd(),
      "/node_modules/@types/nexus-typegen/index.d.ts"
    ),
    schema: path.join(__dirname, "/generated/schema.graphql"),
  },
  contextType: {
    export: "Context",
    module: path.join(__dirname, "/context.ts"),
  },
  sourceTypes: {
    modules: [{ module: ".prisma/client", alias: "PrismaClient" }],
  },
});

export default schema;
