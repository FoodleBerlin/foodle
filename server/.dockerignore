*.png
*.md
*.jpg
#node_modules/ # Some issues appear if this one is not commented out, idk why exactly. TODO: Investigate at later on
tests/