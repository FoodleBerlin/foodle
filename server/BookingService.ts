/* import { Context } from '@apollo/client';
import { Booking, BookingStatus, Frequency } from '@prisma/client';

export function findDaySlots() {}

export function mapRequestedSlotsToExistingSlots() {}

export async function createBooking(
  ctx: Context,
  userId: string,
  propertyId: string,
  price: number,
  startDate: Date,
  endDate: Date,
  frequency: Frequency
): Booking {
  return await ctx.prisma.booking.create({
    data: {
      tenantId: userId,
      propertyId: propertyId,
      bookingStatus: BookingStatus.pending,
      totalPrice: price,
      startDate: startDate,
      endDate: endDate,
      frequency: frequency,
    },
  });
}
 */
