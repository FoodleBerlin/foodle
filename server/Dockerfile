FROM node:16
EXPOSE 5000
RUN apt-get update && apt-get install -y apt-transport-https ca-certificates curl gnupg && \
    curl -sLf --retry 3 --tlsv1.2 --proto "=https" 'https://packages.doppler.com/public/cli/gpg.DE2A7741A397C129.key' | apt-key add - && \
    echo "deb https://packages.doppler.com/public/cli/deb/debian any-version main" | tee /etc/apt/sources.list.d/doppler-cli.list && \
    apt-get update && \
    apt-get -y install doppler

#setup yarn 
RUN npm i -g corepack
RUN corepack prepare yarn@1.22.19 --activate
WORKDIR /
COPY ./ ./server
WORKDIR /server
RUN yarn
RUN yarn prisma:generate
RUN yarn tsc --build
CMD [ "doppler", "run", "--","node", "build/index.js" ]