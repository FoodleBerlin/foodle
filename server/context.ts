import prisma from "./singletons/prisma";
import { PrismaClient } from "@prisma/client";
import { Request, Response } from "express";

export type Context = {
  req: any;
  res: any;
  user: any | null;
  prisma: PrismaClient;
};

export async function createContext(
  req: Request,
  res: Response
): Promise<Context> {
  return {
    req: res,
    res: req,
    user: null, // TODO get is from req.cookies.jwt,
    prisma,
  };
}
