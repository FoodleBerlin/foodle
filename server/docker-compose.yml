version: "3.8"
services:
  # postgresql://foodle:foodle@localhost:5431
  postgres:
    image: postgres
    container_name: foodle
    restart: always
    environment:
      - POSTGRES_USER=foodle
      - POSTGRES_PASSWORD=foodle
    ports:
      - "5431:5432"
