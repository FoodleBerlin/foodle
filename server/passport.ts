const GoogleStrategy = require('passport-google-oauth20');
import passport from 'passport';
import prisma from '../server/singletons/prisma';
import datasources from './singletons/datasources';

const { stripeWrapper } = datasources();

passport.use(
  new GoogleStrategy(
    {
      clientID: process.env.GOOGLE_CLIENT_ID,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
      callbackURL: 'http://localhost:5000/api/callback',
      state: true,
    },
    async function (accessToken: unknown, refreshToken: unknown, profile: any, cb: any) {
      try {
        if (!accessToken || !profile) return cb('error', null);
        const user = await prisma.user.findUnique({
          where: {
            email: profile?.emails[0].value ?? '',
          },
        });
        console.log('user first');
        if (!user) {
          // Create user for them on stripe
          const stripeId = await stripeWrapper.createCustomer({ email: profile.emails[0].value });
          const res = await prisma.user.create({
            data: {
              kind: 'user',
              stripeId: 'cus_Kza1oi2OTlvcp', // hardcoded for now so theres datastripeId.response.success?.body.id,
              handle: profile.emails[0].value,
              fullName: profile.displayName,
              email: profile.emails[0].value,
              image: profile.picture,
              zip: -1,
              description: 'I am from ...',
              isVerified: false,
              role: 'tenant',
              bookings: { create: [] },
              rentingOut: { create: [] },
              superOwner: false,
              passport: 'passportString',
              passportVerified: false,
              license: 'licenseString',
              licenseVerified: false,
              solvency: 'solvencyString',
              solvencyVerified: false,
            },
          });
          console.log('user');
          return cb(null, res);
        } else {
          console.log('cb');
          return cb(null, user);
        }
        console.log('here');
      } catch (e) {
        console.log('ERROR', e);
        return cb(e);
      }
    }
  )
);

passport.serializeUser((user, done) => {
  process.nextTick(function () {
    done(null, { id: 'testyes' });
  });
});
passport.deserializeUser((id, done) => {
  process.nextTick(function () {
    return done(null, { id: 'testyes' });
  });
});

export default passport;
