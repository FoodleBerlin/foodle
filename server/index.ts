import { ApolloServer } from 'apollo-server-express';
import schema from './schema';
import { createContext } from './context';
import express from 'express';
import session from 'express-session';
import passport from './passport';
import forgeJWT from '../utils/forgeJWT';
import { DataSources } from 'apollo-server-core/dist/graphqlOptions';
import StripeWrapper from './singletons/stripe/endpoints';
import datasources from './singletons/datasources';

const port = process.env.PORT || 5000

export const app = express();
app.use(passport.initialize());

app.use(
  session({
    secret: process.env.SERVER_SECRET ?? '',
    resave: false,
    saveUninitialized: false,
  })
);
export const apollo: ApolloServer = new ApolloServer({
  // An executable GraphQL schema.
  schema,
  // An object (or a function that creates an object) that's passed to every resolver that executes for a particular operation.
  introspection: true,
  // This enables resolvers to share helpful context, such as a database connection.
  context: createContext,
  dataSources: () => datasources() as DataSources<Record<'stripeWrapper', StripeWrapper>>,
});

export const router = express.Router();

export async function main() {
  await apollo.start();
  app.use(router);
  apollo.applyMiddleware({ app });
  app.listen({
    port: port,
  });
}
if (!process.env.TEST) {
  main();
}

router.get('/api/auth', passport.authenticate('google', { scope: ['profile', 'email'] }));

router.get('/api/callback', (req: any, res: any, next) => {
  passport.authenticate('google', async (err: any, user: any) => {
    const token = await forgeJWT(user);
    res.cookie('jwt', token, {
      httpOnly: true,
      secure: false, // true in prod,
      sameSite: 'lax', // 'strict' in prod,
    });
    return res.redirect(process.env.CLIENT_URL);
  })(req, res, next);
});
