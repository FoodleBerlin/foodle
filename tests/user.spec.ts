import { apollo as server } from "../server/index";
beforeAll(async () => {
  // TODO reset()
  // TODO seed
});

describe("Find user query", () => {
  const query = `query Query($handle: String!) {
 findUser(handle: $handle) {
   User {
     id
     email
     handle
     fullName
     zip
   }
   ClientErrorUserNotExists {
     message
   }
   ClientErrorInvalidHandle {
     message
   }
 }`;
  it("can query a single user by the handle", async () => {
    const res = await server.executeOperation({
      query,
      variables: { handle: "user1" }, //TODO get this from globals
    });
    expect(res).toMatchSnapshot();
  });
});
