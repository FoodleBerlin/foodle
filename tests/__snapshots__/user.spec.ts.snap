// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Find user query can query a single user by the handle 1`] = `
Object {
  "data": Object {
    "findUser": Object {
      "ClientErrorInvalidHandle": null,
      "ClientErrorUserNotExists": null,
      "User": Object {
        "email": "user1@gmail.com",
        "fullName": "User 1",
        "handle": "user1",
        "zip": 13000,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
